File: backend/app/optical/kinds/mirror/physics.py
"""Mirror PhysicsOp (Python mirror of frontend kinds/mirror/physics.ts).

Op name: reflect_specular
Kind: mirror
Required: face_in.normal_body_local
"""

from __future__ import annotations

from app.optical.beam_ray import BeamRay, Vec3
from app.optical.jones import rotate_jones_into_new_frame
from app.optical.registry import (
    KindEntry,
    PhysicsOpContext,
    register_kind,
)


def reflect_direction(d: Vec3, n: Vec3) -> Vec3:
    """d' = d - 2 (d·n) n"""
    dot_dn = d.dot(n)
    return d - n * (2.0 * dot_dn)


def reflect_specular_op(ray_in: BeamRay, ctx: PhysicsOpContext) -> list[BeamRay]:
    n = ctx.face_in.normal_body_local
    if n is None:
        raise ValueError("reflect_specular: face_in.normal_body_local is required")

    d_new = reflect_direction(ray_in.direction, n)
    reflectivity = float(ctx.params.get("reflectivity", 1.0))
    new_jones = rotate_jones_into_new_frame(ray_in.jones, ray_in.direction, d_new)

    return [ray_in.replaced(
        direction=d_new,
        jones=new_jones,
        power_mw=ray_in.power_mw * reflectivity,
    )]


register_kind("mirror", KindEntry(
    ops={"reflect_specular": reflect_specular_op},
    needs_aperture=True,
))

File: frontend/src/optical/kinds/mirror/physics.ts
/**
 * Mirror PhysicsOp — specular reflection.
 *
 * Op name: `reflect_specular`
 * Kind:    `mirror`
 *
 * Behavior:
 *   - Direction reflects:  d' = d - 2 (d·n) n     (n = face normal, in same frame as d)
 *   - Power × reflectivity (default 1.0)
 *   - Jones basis rotates from old s/p frame into new (post-reflection) s/p frame
 *   - q-parameter unchanged (flat mirror, ideal)
 *   - Origin unchanged (ray stays at face position; next propagation step moves it)
 *
 * Required face: `faceIn.normalBodyLocal` (outward normal, pointing AWAY from
 * the reflective surface — incoming ray has d · n < 0).
 *
 * For Phase 1 the op assumes `faceIn.normalBodyLocal` is expressed in the
 * same frame as `rayIn.direction`. The ray tracer wraps the body↔lab
 * transform later.
 */

import {
  type BeamRay,
  type Vec3,
  vec3Dot,
  vec3Scale,
  vec3Sub,
} from "../../beam-ray";
import { rotateJonesIntoNewFrame } from "../../jones";
import {
  type PhysicsOp,
  type PhysicsOpContext,
  registerKind,
} from "../../registry";

export function reflectDirection(d: Vec3, n: Vec3): Vec3 {
  const dotDN = vec3Dot(d, n);
  return vec3Sub(d, vec3Scale(n, 2 * dotDN));
}

export const reflectSpecularOp: PhysicsOp = (
  rayIn: BeamRay,
  ctx: PhysicsOpContext,
): BeamRay[] => {
  const n = ctx.faceIn.normalBodyLocal;
  if (!n) {
    throw new Error("reflect_specular: faceIn.normalBodyLocal is required");
  }

  const dNew = reflectDirection(rayIn.direction, n);
  const reflectivity = (ctx.params.reflectivity as number | undefined) ?? 1.0;
  const newJones = rotateJonesIntoNewFrame(rayIn.jones, rayIn.direction, dNew);

  return [{
    ...rayIn,
    direction: dNew,
    jones: newJones,
    powerMw: rayIn.powerMw * reflectivity,
    // origin, qx, qy, wavelength, pathLengthMm: unchanged at the reflection point
  }];
};

registerKind("mirror", {
  ops: {
    reflect_specular: reflectSpecularOp,
  },
  needsAperture: true,
});
